### PR TITLE
GetPeerSubscriptions RPC call and test

### PR DESCRIPTION
**Pull request by holisticode**
This PR had originally been opened as https://github.com/ethereum/go-ethereum/pull/18284 but closed here: https://github.com/ethereum/go-ethereum/pull/18284#issuecomment-448640129. It had to be closed because it was using the old `kad.EachBin` which was being refactored in a different PR.

This PR is a iteration of the former PR and is an adaptation to the new subscription code in `stream` and `kademlia`.

This PR adds to the debugging information available.

Over RPC, this code allows to request the stream syncing subscriptions and returns them as a map.
It adds a simulation test which over a snapshot requests all subscriptions.
If provided with a -printstats true flag, it will print the results to STDOUT

### CODE REVIEW

File: swarm/network/stream/streamer_test.go [hunk 1]
```diff
@@ -1114,3 +1121,170 @@ func TestRequestPeerSubscriptions(t *testing.T) {
...
+					// we need to ignore those, we are only counting SYNC streams
+					if !strings.HasPrefix(s, "RETRIEVE_REQUEST") {
+						realCount++
+					}
+				}
+			}
+		}
+		// every node is mutually subscribed to each other, so the actual count is half of it
+		if realCount/2 != expectedMsgCount {
+			return errors.New(fmt.Sprintf("Real subscriptions and expected amount don't match; real: %d, expected: %d", realCount/2, expectedMsgCount))
```
Comment (janos) [inline swarm/network/stream/streamer_test.go]: Use fmt.Errorf.

File: swarm/network/stream/streamer_test.go [hunk 2]
```diff
@@ -1114,3 +1121,170 @@ func TestRequestPeerSubscriptions(t *testing.T) {
...
+			}
+
+			//ask it for subscriptions
+			pstreams := make(map[string][]string)
+			err = client.Call(&pstreams, "stream_getPeerSubscriptions")
+			if err != nil {
+				t.Fatal(err)
+			}
+			//length of the subscriptions can not be smaller than number of peers
+			log.Debug(fmt.Sprintf("node %s subscriptions:", node.String()))
```
Comment (nonsense) [inline swarm/network/stream/streamer_test.go]: `log.Debug` accepts structured log messages, no need for `fmt.Sprintf`

File: swarm/network/stream/streamer_test.go [hunk 3]
```diff
@@ -1114,3 +1121,170 @@ func TestRequestPeerSubscriptions(t *testing.T) {
...
+			for p, ps := range pstreams {
+				log.Debug(fmt.Sprintf("...with node %s: ", p))
```
Comment (nonsense) [inline swarm/network/stream/streamer_test.go]: log.Debug accepts structured log messages, no need for fmt.Sprintf

File: swarm/network/stream/streamer_test.go [hunk 4]
```diff
@@ -1114,3 +1121,170 @@ func TestRequestPeerSubscriptions(t *testing.T) {
...
+				for _, s := range ps {
+					log.Debug(fmt.Sprintf("......%s", s))
```
Comment (nonsense) [inline swarm/network/stream/streamer_test.go]: log.Debug accepts structured log messages, no need for fmt.Sprintf

File: swarm/network/stream/streamer_test.go (hunk 1 above)
Comment (nonsense) [inline swarm/network/stream/streamer_test.go]: You could use `fmt.Errorf` which is simpler than errors.New(fmt.Sprintf(...))

File: swarm/network/stream/streamer_test.go [hunk 5]
```diff
@@ -1114,3 +1121,170 @@ func TestRequestPeerSubscriptions(t *testing.T) {
...
+		nodes := sim.Net.Nodes
+
+		//wait until all subscriptions are done
+		select {
+		case <-allSubscriptionsDone:
+		case <-ctx.Done():
+			t.Fatal("Context timed out")
+		}
+
+		log.Debug("Expected message count: ", "expectedMsgCount", expectedMsgCount)
```
Comment (nonsense) [inline swarm/network/stream/streamer_test.go]: No need for ": ", just:

```
log.Debug("Expected message count", "count", expectedMsgCount)
```

File: swarm/network/stream/streamer_test.go [hunk 6]
```diff
@@ -1114,3 +1121,170 @@ func TestRequestPeerSubscriptions(t *testing.T) {
 		}
 	}
 }
+
+/*
+TestGetSubscriptionsRPC sets up a simulation network of 16 nodes,
+starts the simulation, waits for SyncUpdateDelay in order to kick off
+stream registration, then tests that there are subscriptions.
+*/
+func TestGetSubscriptionsRPC(t *testing.T) {
```
Comment (nonsense) [inline swarm/network/stream/streamer_test.go]: I'd rather we remove this test, provided that subscriptions are tested elsewhere. The RPC API just iterates over all subscriptions and returns them, I don't think adding a simulation test that takes 2sec. on my machine brings much value.

I think this is actually introducing technical debt by adding a simulation test here.

File: swarm/network/stream/streamer_test.go [hunk 7]
```diff
@@ -1265,11 +1322,11 @@ func TestGetSubscriptionsRPC(t *testing.T) {
 				t.Fatal(err)
 			}
 			//length of the subscriptions can not be smaller than number of peers
-			log.Debug(fmt.Sprintf("node %s subscriptions:", node.String()))
+			log.Debug("node subscriptions:", "node", node.String())
```
Comment (nonsense) [inline swarm/network/stream/streamer_test.go]: The `String()` method should be called automatically, you don't have to explicitly call it.

**Pull request merged**